SignalServiceKit/src/Messages/OWSBatchMessageProcessor.m

```diff
@@ -300,32 +326,35 @@ - (void)drainQueueWorkStep
         return;
     }
 
-    [self processJobs:jobs
-           completion:^{
-               dispatch_async(dispatch_get_main_queue(), ^{
-                   [self.finder removeJobsWithIds:jobs.uniqueIds];
-
-                   DDLogVerbose(@"%@ completed %zd jobs. %zd jobs left.",
-                       self.tag,
-                       jobs.count,
-                       [OWSMessageContentJob numberOfKeysInCollection]);
-                   [self drainQueueWorkStep];
-               });
-           }];
+    [self processJobs:jobs];
+
+    [self.finder removeJobsWithIds:jobs.uniqueIds];
+
+    DDLogVerbose(@"%@ completed %zd jobs. %zd jobs left.",
+        self.tag,
+        jobs.count,
+        [OWSMessageContentJob numberOfKeysInCollection]);
+
+    // Wait a bit in hopes of increasing the batch size.
+    // This delay won't affect the first message to arrive when this queue is idle,
+    // so by definition we're receiving more than one message and can benefit from
+    // batching.
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1f * NSEC_PER_SEC)), self.serialQueue, ^{
```

Comment:
With a delay of 0.1 seconds, we tend to get batch sizes of 1-2.  Larger batches would be better, but decryption is much more expensive than message processing, so the message processing pipeline never accumulates a larger batch.